### PR TITLE
Change stringifying options behavior

### LINKED ISSUE #174: JSON option when posting data
Would it be possible to use the `json: true` option when posting data to set `Content-Type` to `application/json` and stringify the body with `JSON.stringify`?

### PR DESCRIPTION
**Pull request by AlexTes**
Changes the behavior of the stringifying options.

* [x] don't accept plain object unless `form` or `json` is true
* [x] stringify if `json` is true
* [x] stringify if `form` is true
* [x] stringify as querystring and parse `res` as JSON if `form` and `json` are true

Things I changed because I found the code hard to deal with, and confusing without.
* clean up input validation using Sindre's suggestion from #212.
* simpler use of `body` variable. It's a shorthand for the body passed by the user. `opts.body` is what we mutate into the body we want to send.
* 'throws on bad input' test is maybe overkill

As always details are in the commit messages.

Sidenotes: I went for minimal change based on the proposal from #174. I think this can be quite a bit cleaner still. I'm considering a second PR. Specifically what I feel could be better:
* make 'body is string' case explicit instead of implicit fallback
* use early returns to limit code paths and sharpen reduce redundant input checking
* some tests use destructuring others don't _for the exact same response property_.

---
Fixes #174

### CODE REVIEW

**Comment by sindresorhus** [comment]
@floatdrop What do you think of these changes? See my proposal in https://github.com/sindresorhus/got/issues/174#issuecomment-298292987

**Comment by sindresorhus** [comment]
> I'm considering a second PR. Specifically what I feel could be better:

:+1:

**Comment by AlexTes** [comment]
I got an avocado rainbow 😲 ✊!

File: index.js [hunk 1]
```diff
@@ -236,31 +237,37 @@ function normalizeArguments(url, opts) {
...
-			throw new Error('options.body must be a ReadableStream, string, Buffer or plain Object');
+	const body = opts.body;
+	if (body !== null && body !== undefined) {
+		const headers = opts.headers;
+		if (!isStream(body) && typeof body !== 'string' && !Buffer.isBuffer(body) && !(opts.form || opts.json)) {
+			throw new TypeError('options.body must be a ReadableStream, string, Buffer or plain Object');
 		}
 
-		opts.method = opts.method || 'POST';
+		if ((opts.form || opts.json) && !(isObj(body) && typeof body !== 'function')) {
```
Comment (floatdrop) [inline index.js]: Is `typeof body !== 'funciton'` is necessary? Looks like `isPlainObj(function() {})` returns `false`.

File: index.js [hunk 2]
```diff
@@ -236,31 +237,37 @@ function normalizeArguments(url, opts) {
...
+			throw new TypeError('options.body must be a plain Object');
```
Comment (floatdrop) [inline index.js]: `options.body must be a plain Object when options.form or options.json is used`

File: index.js [hunk 3]
```diff
@@ -236,31 +237,37 @@ function normalizeArguments(url, opts) {
...
-			const length = typeof body === 'string' ? Buffer.byteLength(body) : body.length;
-			opts.headers['content-length'] = length;
+		if (headers['content-length'] === undefined && headers['transfer-encoding'] === undefined && !isStream(body)) {
+			const length = typeof opts.body === 'string' ? Buffer.byteLength(opts.body) : opts.body.length;
+			headers['content-length'] = length;
 		}
-	}
 
-	opts.method = (opts.method || 'GET').toUpperCase();
+		opts.method = opts.method || 'POST';
```
Comment (floatdrop) [inline index.js]: `opts.method` should be uppercased in this if-branch too.

File: index.js (hunk 1 above)
Comment (AlexTes) [inline index.js]: Nice catch. I was confusing `is-obj` and `is-plain-obj`.

**Comment by sindresorhus** [comment]
Thank you for yet another excellent pull request @AlexTes. I'd like to invite you to join the project if you're interested? 🙏

**Comment by AlexTes** [comment]
@sindresorhus 🙇 🙇 . You just made my day. And it's only 10 AM!

**Pull request merged**